### ISSUE TITLE
Generate a `.XCompose` file

### ISSUE THREAD

**Issue by lawcho** [agda/agda#7083]
I'd like to use `agda-mode`'s unicode key-bindings system-wide.

This is possible by writing a `.XCompose` file containing key-bindings [^1] [^2] (or generating one [^3])

To avoid duplicated effort, the tooling in this repo could generate a `.XCompose` file
from the same source as `agda-mode`:

![agda-input-mode-new-arch excalidraw](https://github.com/agda/agda/assets/95857153/249e666a-415d-422c-9f92-43d48bd9d17a)

Programming tasks required to support this:

* **Emacs Lisp**: update [`agda-input.el`][] to load key-bindings from a JSON file
* **sed**: port the big list of bindings from [`agda-input.el`][] to a JSON file
* **Haskell**: write a JSON → `.XCompose` generator script
* **GH actions**: distribute the `.XCompose` file via CI/CD
* **ReStructuredText**: document in the [user manual][]

[^1]: https://github.com/useronym/dotfiles/blob/198e/.XCompose
[^2]: https://github.com/noughtmare/haskell-xcompose/blob/77ceb/.XCompose
[^3]: https://github.com/np/xcompose/blob/3d80/NP/Unicode.hs

[`agda-input.el`]: https://github.com/agda/agda/blob/master/src/data/emacs-mode/agda-input.el

[user manual]: https://github.com/agda/agda/blob/master/doc/user-manual/tools/emacs-mode.rst#unicode-input

**Comment by lawcho** [comment]
Related (but `LaTeX`-oriented) discussion in #3645

**Comment by nad** [comment]
Note that the Agda input method (by default) inherits some key bindings from Emacs' TeX input method:

https://github.com/agda/agda/blob/fbf9d159c3c874b8328ccdc78a0d57d57a310234/src/data/emacs-mode/agda-input.el#L150-L175

If you dump those key bindings to a file and use them in some project, then it might make sense for that project to use the GPL.

**Comment by jul1u5** [comment]
Note that, at least for me on GNOME (Wayland) + ibus, I cannot have multiple key-bindings in `.XCompose` with a common prefix. For example, take the following key-bindings (which are defined in agda-input)
```
<-	←
<->	↔
```
In `.XCompose` format, this would be written as:
```
<Multi_key> <less> <minus>				: "←"
<Multi_key> <less> <minus> <greater>	: "↔"
```
Now, if I press the compose key, <kbd>\<</kbd>, and <kbd>-</kbd> I get `←`, which is correct. But as soon as I press <kbd>-</kbd>, ibus finds a key-binding that matches and closes the "compose mode". So, when trying to type `↔` (compose key, <kbd>\<</kbd>, <kbd>-</kbd>, <kbd>></kbd>), I get `←>` instead.

I don't know if other input methods can handle such key-bindings.

One way to make this work, would be to use a "termination" character to "submit" a key-binding, so that we can delay the matching behavior until we type the whole key-binding. For instance, the following key-bindings allows me to type both symbols:
```
<Multi_key> <less> <minus> <period>				: "←"
<Multi_key> <less> <minus> <greater> <period>	: "↔"
```

**Comment by lawcho** [comment]
Yes, I'd expect the `.XCompose` generator to insert characters for...

* *termination* -- e.g. [np's `.XCompose` generator][] inserts both `<space>` and `<tab>` terminators for prefixes:

        ...
        <Multi_key> <c> <space> : "⊂"
        <Multi_key> <c> <tab> : "⊂"
        <Multi_key> <slash> <c> <space> : "⊄"
        <Multi_key> <slash> <c> <tab> : "⊄"
        <Multi_key> <c> <minus> <space> : "∈"
        <Multi_key> <c> <minus> <tab> : "∈"
        <Multi_key> <slash> <c> <minus> : "∉"
        <Multi_key> <c> <slash> <equal> : "⊊"
        <Multi_key> <r> <c> <equal> : "⊇"
        <Multi_key> <r> <c> <space> : "⊃"
        <Multi_key> <r> <c> <tab> : "⊃"
        ...


* *navigation* -- e.g. to distinguish between the 4 symbols that `(A)` maps to:

        <Multi_key> <parenleft> <A> <parenright> <space> : "Ⓐ"
        <Multi_key> <parenleft> <A> <parenright> <Right> <space> : "ⓐ"
        <Multi_key> <parenleft> <A> <parenright> <Right> <Right> <space> : "🅐"
        <Multi_key> <parenleft> <A> <parenright> <Right> <Right> <Right> <space> : "🄰"
        <Multi_key> <parenleft> <A> <parenright> <Right> <Right> <Right> <Right> <space> : "🅰"

Both of these be configurable via generator-script args.

[np's `.XCompose` generator]: https://github.com/np/xcompose/blob/3d80/NP/Unicode.hs

**Comment by nad** [comment]
This pull request seems to be incompatible with pull request #6538.

**Comment by lawcho** [comment]
Yes, #7094 is incompatible with #6538 (and the TeX-inherited bindings).

Here are 3 ways to get compatability:

### Only elisp

![](https://github.com/agda/agda/assets/95857153/b7ece2be-02e3-44e6-957e-750a150e7305)

### Minimum elisp

![](https://github.com/agda/agda/assets/95857153/2f1f5a62-ab63-4951-9da6-0efbe2454b04)

### Mixed

![](https://github.com/agda/agda/assets/95857153/7be96c57-acc5-4361-82f8-6f3a68c4047d)

## Notes

I see a trade-off between:

* Amount of `.el` code to maintain
* Build complexity of `agda-input.el`
* Build complexity of `.XCompose`
